src/tools/tidy/src/rustdoc_html_ids.rs

```diff
@@ -0,0 +1,143 @@
+//! Checks that all error codes have at least one test to prevent having error
+//! codes that are silently not thrown by the compiler anymore.
+
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use regex::Regex;
+
+const ID_MAP_PATH: &str = "librustdoc/html/markdown.rs";
+const IDS_TO_IGNORE: &[&str] = &[
+    // This one is created in the JS cannot be found in rust files.
```

Comment:
Shouldn't we rather say that this should *not* be found in the Rust files, rather than being ignored? If Rust starts generating this ID, that would be a problem, right?

Reply:
Doesn't work this way. ;)

The problem here can come from multiple places: the IDs generated for each item on doc pages and the "manual" HTML code generated by users from the markdown (the titles generate IDs). So in here, we declare an ID that isn't used the rust source code and we tell the linter to not throw an error for an unused ID.

Reply:
I think I follow both of those locations, but this comment suggests a *third* location where IDs are generated -- the JS code emitted by rustdoc, specifically https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/static/main.js#L893 ?

We declare that ID appropriately in the rustdoc Rust map (https://github.com/rust-lang/rust/pull/86178/files#diff-9797963bcc8ee935b3ce76a245b60bed2d5045deb1f170accb86c14c88b27908R1361), but it shouldn't be *ignored* there -- it should be checked to be **only** externally generated. That is, we want the Rust code to never emit that ID, right?

Reply:
Sorry, my explanation was bad I guess. There are two errors:

1. If an ID is used but isn't declared in the ID map.
2. If an ID is declared in the ID map but isn't used **in the rust code**.

This ID enters the second category because it is created in the JS source code. We need to tell the linter to not emit an error for an unused ID because it doesn't find it in the rust source code.

Reply:
We shouldn't call that *ignore* though -- it should be explicitly *not* in the Rust code, right? If the rust code starts to use it, then in all likelihood the JS will break.

Reply:
Ah, yes, the name is pretty bad now that I think about it...

And yes, it's a very good point! I need to check that the "ignored" (really needs another name, suggestions open!) are not used in the rust code. Great catch!

Reply:
I renamed it `IDS_USED_IN_JS`.

Reply:
Just because JS uses the ID, doesn't mean it needs to go in this list though, right? Only IDs which are exclusively created by JS need to be placed in this list.

It would also be helpful for the comment -- which *I think* should be broadly applicable, not just to 'help', to be moved outside the array and onto the const itself.

Reply:
No, the JS *creates* this ID, it doesn't just use it, which is why it is declared here.